fendermint/vm/topdown/src/finality.rs

```diff
@@ -0,0 +1,479 @@
+    fn validator_set(&self, height: BlockHeight) -> StmResult<Option<ValidatorSet>, Error> {
+        let cache = self.height_data.read()?;
+        Ok(cache.get_value(height).map(|i| i.1.clone()))
+    }
+
+    fn top_down_msgs(
+        &self,
+        from_height: BlockHeight,
+        to_height: BlockHeight,
+    ) -> StmResult<Vec<CrossMsg>, Error> {
+        let cache = self.height_data.read()?;
+        let v = cache
+            .values_within(from_height, to_height)
+            .flat_map(|i| i.2.iter())
+            .cloned()
+            .collect();
+        Ok(v)
+    }
+}
+
+impl ParentViewProvider for InMemoryFinalityProvider {
+    fn latest_height(&self) -> StmResult<Option<BlockHeight>, Error> {
+        let h = self.parent_view_data.latest_height()?;
+        Ok(h)
+    }
+
+    fn block_hash(&self, height: BlockHeight) -> StmResult<Option<BlockHash>, Error> {
+        let v = self.parent_view_data.block_hash(height)?;
+        Ok(v)
+    }
+
+    fn validator_set(&self, height: BlockHeight) -> StmResult<Option<ValidatorSet>, Error> {
+        let v = self.parent_view_data.validator_set(height)?;
+        Ok(v)
+    }
+
+    fn top_down_msgs(&self, height: BlockHeight) -> StmResult<Vec<CrossMsg>, Error> {
+        let v = self.parent_view_data.top_down_msgs(height, height)?;
+        Ok(v)
+    }
+
+    fn new_parent_view(
+        &self,
+        height: BlockHeight,
+        block_hash: BlockHash,
+        validator_set: ValidatorSet,
+        top_down_msgs: Vec<CrossMsg>,
+    ) -> StmResult<(), Error> {
+        if !top_down_msgs.is_empty() {
+            // make sure incoming top down messages are ordered by nonce sequentially
+            ensure_sequential_by_nonce(&top_down_msgs)?;
+        };
+
+        let r = self.parent_view_data.height_data.modify(|mut cache| {
+            let r = cache
+                .append(height, (block_hash, validator_set, top_down_msgs))
+                .map_err(Error::NonSequentialParentViewInsert);
+            (cache, r)
+        })?;
+
+        if let Err(e) = r {
+            return abort(e);
+        }
+
+        Ok(())
+    }
+}
+
+impl ParentFinalityProvider for InMemoryFinalityProvider {
+    fn last_committed_finality(&self) -> StmResult<IPCParentFinality, Error> {
+        let finality = self
+            .last_committed_finality
+            .read_clone()?
+            .ok_or(StmError::Abort(Error::CommittedFinalityNotReady))?;
+        Ok(finality)
+    }
+
+    fn next_proposal(&self) -> StmResult<Option<IPCParentFinality>, Error> {
+        let latest_height = if let Some(h) = self.parent_view_data.latest_height()? {
+            h
+        } else {
+            return abort(Error::HeightNotReady);
+        };
+
+        // latest height has not reached, we should wait or abort
+        if latest_height < self.config.chain_head_delay {
+            return abort(Error::HeightThresholdNotReached);
+        }
+
+        let height = latest_height - self.config.chain_head_delay;
```

Comment:
This seems to be not needed any more, because the syncer only adds data that is final (or should only do so anyway, like I said you should add docstrings to communicate this).

Reply:
Really nice catch. I'm doing the integration testing, and with all the updates, I think the initial logic might have to update a little.

Reply:
Hm, I thought this would change.